File: MaxContiguousSubsequence/MaxContiguousSubsequence.py

def split_and_solve(nums):
    if len(nums) == 1:
        return nums[0]
    middle = len(nums) // 2
    return max(split_and_solve(nums[:middle]), split_and_solve(nums[middle:]), solve_middle(nums, middle))


def solve_middle(nums, middle):
    current_sum = 0
    right_max_subseq = float('-inf')
    for num in nums[middle:]:
        current_sum += num
        if right_max_subseq < current_sum:
            right_max_subseq = current_sum
    current_sum = 0
    left_max_subseq = float('-inf')
    for num in nums[:middle + 1][::-1]:
        current_sum += num
        if left_max_subseq < current_sum:
            left_max_subseq = current_sum
    return max(left_max_subseq + right_max_subseq - nums[middle], left_max_subseq, right_max_subseq)


def divide_and_conquer(file_path):
    nums = [int(a) for a in open(file_path).read()[1:-1].split(', ')]
    return split_and_solve(nums)


def brute_force(file_path):
    nums = [int(a) for a in open(file_path).read()[1:-1].split(', ')]
    max_contig_subseq = 0
    for i in range(len(nums)):
        for j in range(i + 1, len(nums) + 1):
            current_sum = sum(nums[i:j])
            if current_sum > max_contig_subseq:
                max_contig_subseq = current_sum
    return max_contig_subseq


def kadane(file_path):
    nums = [int(a) for a in open(file_path).read()[1:-1].split(', ')]
    current_sum = 0
    max_contig_subseq = float('-inf')
    for num in nums:
        current_sum = max(num, current_sum + num)
        max_contig_subseq = max(current_sum, max_contig_subseq)
    return max_contig_subseq

File: MaxContiguousSubsequence/MaxContiguousSubsequence.cpp
#include <vector>
#include <string>
#include <fstream>
#include <algorithm>
#include "MaxContiguousSubsequence.h"

using namespace std;

vector<int> MaxContiguousSubsequence::parseInput(const string &filePath) {
    vector<int> nums;
    ifstream file(filePath);
    string input;
    while (getline(file, input, ',')) {
        nums.push_back(stoi(input.substr(1)));
    }
    return nums;
}

int MaxContiguousSubsequence::kadane(const string &filePath) {
    ifstream file(filePath);
    string input;
    int currentSum = 0;
    int maxContigSubseq = 0;
    while (getline(file, input, ',')) {
        int v = stoi(input.substr(1));
        currentSum = max(v, currentSum + v);
        if (currentSum > maxContigSubseq) {
            maxContigSubseq = currentSum;
        }
    }
    return maxContigSubseq;
}

int MaxContiguousSubsequence::splitAndSolve(int* nums, int numsCount) {
    if (numsCount == 1) {
        return nums[0];
    }
    int middle = numsCount / 2;
    int* left = nums;
    int* right = nums + middle;
    return max(max(splitAndSolve(left, middle), splitAndSolve(right, middle)), solveMiddle(nums, numsCount, middle));
}

int MaxContiguousSubsequence::solveMiddle(int* nums, int numsCount, int middle) {
    int currentSum = 0;
    int rightMaxSubseq = min_element(nums, nums + numsCount)[0];
    for (int i = middle; i < numsCount; ++i) {
        currentSum += nums[i];
        if (rightMaxSubseq < currentSum) {
            rightMaxSubseq = currentSum;
        }
    }
    currentSum = 0;
    int leftMaxSubseq = min_element(nums, nums + numsCount)[0];
    for (int i = middle; i > -1; --i) {
        currentSum += nums[i];
        if (leftMaxSubseq < currentSum) {
            leftMaxSubseq = currentSum;
        }
    }
    return max(max(leftMaxSubseq, rightMaxSubseq), leftMaxSubseq + rightMaxSubseq - nums[middle]);
}

int MaxContiguousSubsequence::divideAndConquer(const string &filePath) {
    vector<int> numsInput = parseInput(filePath);
    int numsCount = int(numsInput.size());
    int nums[numsCount];
    for (int i = 0; i < numsCount; ++i) {
        nums[i] = numsInput[i];
    }
    return splitAndSolve(nums, numsCount);
}

int MaxContiguousSubsequence::bruteForce(const string &filePath) {
    vector<int> numsInput = parseInput(filePath);
    int numsCount = int(numsInput.size());
    int nums[numsCount];
    for (int i = 0; i < numsCount; ++i) {
        nums[i] = numsInput[i];
    }
    int maxContigSubseq = 0;
    for (int i = 0; i < numsCount; i++) {
        for (int j = i + 1; j <= numsCount; j++) {
            int currentSum = 0;
            for (int k = i; k < j; ++k) {
                currentSum += nums[k];
            }
            if (currentSum > maxContigSubseq) {
                maxContigSubseq = currentSum;
            }
        }
    }
    return maxContigSubseq;
}
